File: src/chore_reminder/scheduler.py
from __future__ import annotations

from calendar import monthrange
from dataclasses import dataclass
from datetime import date, datetime, time, timedelta
from zoneinfo import ZoneInfo

from chore_reminder.messages import good_morning_message


@dataclass(frozen=True)
class Occurrence:
    schedule_id: str
    occurrence_index: int
    task_id: str
    assignee_id: str
    assignee_name: str
    phone: str | None
    task: str
    due_at: datetime
    due_by: datetime
    message: str

    @property
    def reminder_id(self) -> str:
        return f"{self.schedule_id}:{self.occurrence_index}"


def due_occurrences(config: dict, now: datetime | None = None) -> list[Occurrence]:
    tz = ZoneInfo(config.get("timezone", "America/Vancouver"))
    now = now.astimezone(tz) if now else datetime.now(tz)
    lookback_days = int(config.get("lookback_days", 14))
    occurrences: list[Occurrence] = []

    for schedule in config["schedules"]:
        start_date = date.fromisoformat(schedule["start_date"])
        interval = schedule["interval"]
        due_window = schedule["due_window"]
        reminder_time = time.fromisoformat(schedule["reminder_time"])

        first_index = max(0, _index_at_or_before(start_date, now.date() - timedelta(days=lookback_days), interval))
        last_index = _index_at_or_before(start_date, now.date(), interval)

        for index in range(first_index, last_index + 1):
            due_date = _add_interval(start_date, interval, index)
            due_at = datetime.combine(due_date, reminder_time, tzinfo=tz)
            due_by = _due_by(due_at, due_window)
            if due_at <= now <= due_by:
                rotation_item = schedule["rotation"][index % len(schedule["rotation"])]
                person = config["people"][rotation_item["assignee"]]
                message_template = schedule.get(
                    "message_template",
                    "{assignee}, today's chore is: {task}. Reply Y when done.",
                )
                task_message = message_template.format(
                    assignee=person["display_name"],
                    task=rotation_item["task"],
                    due_at=due_at.isoformat(),
                    due_by=due_by.isoformat(),
                )
                occurrences.append(
                    Occurrence(
                        schedule_id=schedule["id"],
                        occurrence_index=index,
                        task_id=rotation_item["task_id"],
                        assignee_id=rotation_item["assignee"],
                        assignee_name=person["display_name"],
                        phone=_person_phone(person),
                        task=rotation_item["task"],
                        due_at=due_at,
                        due_by=due_by,
                        message=f"{good_morning_message()} {task_message}",
                    )
                )

    return sorted(occurrences, key=lambda item: item.due_at)


def next_occurrences(config: dict, count: int = 10, now: datetime | None = None) -> list[Occurrence]:
    tz = ZoneInfo(config.get("timezone", "America/Vancouver"))
    now = now.astimezone(tz) if now else datetime.now(tz)
    found: list[Occurrence] = []
    horizon_days = max(400, count * 40)

    for schedule in config["schedules"]:
        start_date = date.fromisoformat(schedule["start_date"])
        interval = schedule["interval"]
        reminder_time = time.fromisoformat(schedule["reminder_time"])
        start_index = max(0, _index_at_or_before(start_date, now.date(), interval))

        for index in range(start_index, start_index + horizon_days):
            due_date = _add_interval(start_date, interval, index)
            due_at = datetime.combine(due_date, reminder_time, tzinfo=tz)
            if due_at <= now:
                continue
            rotation_item = schedule["rotation"][index % len(schedule["rotation"])]
            person = config["people"][rotation_item["assignee"]]
            found.append(
                Occurrence(
                    schedule_id=schedule["id"],
                    occurrence_index=index,
                    task_id=rotation_item["task_id"],
                    assignee_id=rotation_item["assignee"],
                    assignee_name=person["display_name"],
                    phone=_person_phone(person),
                    task=rotation_item["task"],
                    due_at=due_at,
                    due_by=_due_by(due_at, schedule["due_window"]),
                    message="",
                )
            )
            if len(found) >= count:
                break

    return sorted(found, key=lambda item: item.due_at)[:count]


def _person_phone(person: dict) -> str | None:
    phone = person.get("phone")
    return phone if phone else None


def _index_at_or_before(start: date, target: date, interval: dict) -> int:
    if target < start:
        return 0
    every = int(interval["every"])
    if interval["unit"] == "day":
        return (target - start).days // every

    months = (target.year - start.year) * 12 + target.month - start.month
    return max(0, months // every)


def _add_interval(start: date, interval: dict, index: int) -> date:
    every = int(interval["every"])
    if interval["unit"] == "day":
        return start + timedelta(days=every * index)
    return _add_months(start, every * index)


def _due_by(due_at: datetime, due_window: dict) -> datetime:
    amount = int(due_window["amount"])
    if due_window["unit"] == "day":
        return due_at + timedelta(days=amount)
    return datetime.combine(_add_months(due_at.date(), amount), due_at.time(), due_at.tzinfo)


def _add_months(value: date, months: int) -> date:
    month_index = value.month - 1 + months
    year = value.year + month_index // 12
    month = month_index % 12 + 1
    day = min(value.day, _days_in_month(year, month))
    return date(year, month, day)


def _days_in_month(year: int, month: int) -> int:
    return monthrange(year, month)[1]

File: firebase/functions/src/scheduler.ts
import { goodMorningMessage } from "./messages";
import type { AppConfig, Occurrence } from "./types";

const DAY_MS = 24 * 60 * 60 * 1000;

export function dueOccurrences(config: AppConfig, now = new Date()): Occurrence[] {
  const lookbackDays = config.lookbackDays ?? 14;
  const occurrences: Occurrence[] = [];

  for (const schedule of config.schedules) {
    const startDate = parseDate(schedule.startDate);
    const targetDate = localDateInTimezone(now, config.timezone);
    const firstIndex = Math.max(0, indexAtOrBefore(startDate, addDays(targetDate, -lookbackDays), schedule.interval));
    const lastIndex = indexAtOrBefore(startDate, targetDate, schedule.interval);

    for (let index = firstIndex; index <= lastIndex; index += 1) {
      const dueDate = addInterval(startDate, schedule.interval, index);
      const dueAt = zonedDateTime(dueDate, schedule.reminderTime, config.timezone);
      const dueBy = addDueWindow(dueAt, schedule.dueWindow);
      if (dueAt <= now && now <= dueBy) {
        occurrences.push(buildOccurrence(config, schedule, index, dueAt, dueBy));
      }
    }
  }

  return occurrences.sort((left, right) => left.dueAt.getTime() - right.dueAt.getTime());
}

export function occurrencesBetween(config: AppConfig, startsAt: Date, endsAt: Date): Occurrence[] {
  if (endsAt <= startsAt) return [];

  const occurrences: Occurrence[] = [];
  const firstLocalDate = localDateInTimezone(startsAt, config.timezone);
  const lastLocalDate = localDateInTimezone(new Date(endsAt.getTime() - 1), config.timezone);

  for (const schedule of config.schedules) {
    const startDate = parseDate(schedule.startDate);
    const firstIndex = Math.max(0, indexAtOrBefore(startDate, firstLocalDate, schedule.interval) - 1);
    const lastIndex = indexAtOrBefore(startDate, lastLocalDate, schedule.interval) + 1;

    for (let index = firstIndex; index <= lastIndex; index += 1) {
      const dueDate = addInterval(startDate, schedule.interval, index);
      const dueAt = zonedDateTime(dueDate, schedule.reminderTime, config.timezone);
      if (dueAt < startsAt || dueAt >= endsAt) continue;
      const dueBy = addDueWindow(dueAt, schedule.dueWindow);
      occurrences.push(buildOccurrence(config, schedule, index, dueAt, dueBy));
    }
  }

  return occurrences.sort((left, right) => left.dueAt.getTime() - right.dueAt.getTime());
}

export function relativeLocalDayRange(
  now: Date,
  timezone: string,
  startDayOffset: number,
  endDayOffset: number,
): { startsAt: Date; endsAt: Date } {
  const localDate = localDateInTimezone(now, timezone);
  return {
    startsAt: zonedDateTime(addDays(localDate, startDayOffset), "00:00", timezone),
    endsAt: zonedDateTime(addDays(localDate, endDayOffset), "00:00", timezone),
  };
}

function buildOccurrence(
  config: AppConfig,
  schedule: AppConfig["schedules"][number],
  index: number,
  dueAt: Date,
  dueBy: Date,
): Occurrence {
  const startOffset = schedule.startAssignee
    ? schedule.rotation.findIndex((item) => item.assignee === schedule.startAssignee)
    : 0;
  const rotationItem = schedule.rotation[(startOffset + index) % schedule.rotation.length];
  const person = config.people[rotationItem.assignee];
  const taskMessage = (schedule.messageTemplate ?? "{assignee}, today's chore: {task}. Text Y when handled or S to skip.")
    .replaceAll("{assignee}", person.displayName)
    .replaceAll("{task}", rotationItem.task)
    .replaceAll("{due_at}", dueAt.toISOString())
    .replaceAll("{due_by}", dueBy.toISOString());

  return {
    reminderId: `${schedule.id}:${index}`,
    scheduleId: schedule.id,
    occurrenceIndex: index,
    taskId: rotationItem.taskId,
    assigneeId: rotationItem.assignee,
    assigneeName: person.displayName,
    phone: person.phone || undefined,
    task: rotationItem.task,
    dueAt,
    dueBy,
    message: `${goodMorningMessage()} ${taskMessage}`,
  };
}

function parseDate(value: string): Date {
  const [year, month, day] = value.split("-").map(Number);
  return new Date(Date.UTC(year, month - 1, day));
}

function localDateInTimezone(value: Date, timezone: string): Date {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: timezone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).formatToParts(value);
  const year = Number(parts.find((part) => part.type === "year")?.value);
  const month = Number(parts.find((part) => part.type === "month")?.value);
  const day = Number(parts.find((part) => part.type === "day")?.value);
  return new Date(Date.UTC(year, month - 1, day));
}

function zonedDateTime(date: Date, hhmm: string, timezone: string): Date {
  const [hour, minute] = hhmm.split(":").map(Number);
  const utcGuess = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), hour, minute));
  const offsetMinutes = timezoneOffsetMinutes(utcGuess, timezone);
  return new Date(utcGuess.getTime() - offsetMinutes * 60 * 1000);
}

function timezoneOffsetMinutes(value: Date, timezone: string): number {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: timezone,
    timeZoneName: "shortOffset",
  }).formatToParts(value);
  const offset = parts.find((part) => part.type === "timeZoneName")?.value ?? "GMT";
  const match = offset.match(/^GMT([+-])(\d{1,2})(?::(\d{2}))?$/);
  if (!match) return 0;
  const sign = match[1] === "+" ? 1 : -1;
  const hours = Number(match[2]);
  const minutes = Number(match[3] ?? "0");
  return sign * (hours * 60 + minutes);
}

function indexAtOrBefore(start: Date, target: Date, interval: { every: number; unit: "day" | "month" }): number {
  if (target < start) return 0;
  if (interval.unit === "day") {
    return Math.floor((target.getTime() - start.getTime()) / DAY_MS / interval.every);
  }
  const months = (target.getUTCFullYear() - start.getUTCFullYear()) * 12 + target.getUTCMonth() - start.getUTCMonth();
  return Math.max(0, Math.floor(months / interval.every));
}

function addInterval(start: Date, interval: { every: number; unit: "day" | "month" }, index: number): Date {
  if (interval.unit === "day") {
    return addDays(start, interval.every * index);
  }
  return addMonths(start, interval.every * index);
}

function addDueWindow(dueAt: Date, dueWindow: { amount: number; unit: "day" | "month" }): Date {
  if (dueWindow.unit === "day") {
    return new Date(dueAt.getTime() + dueWindow.amount * DAY_MS);
  }
  return addMonths(dueAt, dueWindow.amount);
}

function addDays(value: Date, days: number): Date {
  return new Date(value.getTime() + days * DAY_MS);
}

function addMonths(value: Date, months: number): Date {
  const result = new Date(value);
  const day = result.getUTCDate();
  result.setUTCDate(1);
  result.setUTCMonth(result.getUTCMonth() + months);
  const lastDay = new Date(Date.UTC(result.getUTCFullYear(), result.getUTCMonth() + 1, 0)).getUTCDate();
  result.setUTCDate(Math.min(day, lastDay));
  return result;
}
